sdk/metric/benchmark_test.go

```diff
@@ -43,38 +43,60 @@ func BenchmarkCounterAddNoAttrs(b *testing.B) {
 }
 
 func BenchmarkCounterAddOneAttr(b *testing.B) {
+	opt := []instrument.Int64AddOption{
+		instrument.WithAttributes(attribute.String("K", "V")),
+	}
 	ctx, _, cntr := benchCounter(b)
 
 	for i := 0; i < b.N; i++ {
-		cntr.Add(ctx, 1, attribute.String("K", "V"))
+		cntr.Add(ctx, 1, opt...)
 	}
 }
 
 func BenchmarkCounterAddOneInvalidAttr(b *testing.B) {
+	opt := []instrument.Int64AddOption{
+		instrument.WithAttributes(
+			attribute.String("", "V"),
+			attribute.String("K", "V"),
+		),
+	}
 	ctx, _, cntr := benchCounter(b)
 
 	for i := 0; i < b.N; i++ {
-		cntr.Add(ctx, 1, attribute.String("", "V"), attribute.String("K", "V"))
+		cntr.Add(ctx, 1, opt...)
 	}
 }
 
 func BenchmarkCounterAddSingleUseAttrs(b *testing.B) {
+	attrs := make([]attribute.KeyValue, 1)
+	opt := make([]instrument.Int64AddOption, 1)
```

Comment:
This is not a fair comparison; you can't preallocate both your attributes and options.

This should be testing the more obvious way people will use it.  If you want to include this as an additional optimized benchmark, then I think that would be fair.

Reply:
Why not? If a user is conscious of performance this is indeed the way they will use this option for non-static attributes. We can always add less optimal fixtures to a benchmark that can mask the results from the desired measure behavior. It's just that is the antithesis of what we want to do with a benchmark.

As mentioned in the PR description, I believe this is the original way we should have been measuring the performance of `Add` given it reflects a true benchmark of that method and that method alone. It removes the performance overhead of a fixture, that a user can also make, and shows the performance of the method being benchmarked.

Reply:
We could add a "non-optimized" and an "optimized" benchmark, but that will only tell us the performance of the go compiler in its ability to allocate by looking at the overhead of the "non-optimized". That does not seem needed given we are trying to determine the performance of our code, not the Go compiler.

Reply:
This benchmark represented a general case of how the API was used.  It was a canary for changes exactly like this in seeing things like allocation and runtime.  You can't compare the results meaningfully by going from no preallocating to preallocating as much as possible.  

If a user cannot preallocate in this way, you see that this change increases the number of allocations.

Reply:
> This benchmark represented a general case of how the API was used.

I'm not sure that is the case. This benchmark is not named `BenchmarkCounterGeneralUse` and it exists among many other benchmarks testing the performance of the `Add` method of a counter based on its potentially different behaviors. If this were the general use-case it would seem to me that it would named that way and be distinct in testing this part of the code.

On the other hand, I see a benchmark named "BenchmarkCounterAddSingleUseAttrs" and I expect it to benchmark the counter's `Add` method with single use attributes. I expect it to as best as possible show me the performance of that method alone when it is passed a set of single use attributes. Especially when considered in the context of the other benchmarks it is with that all measure the performance of that method with different input.

If we include allocation of the benchmark fixtures in the benchmark then that benchmark needs to be evaluated with a caveat. We would then have to say "here is the performance of the Add method when passed a single use attribute, **but** it includes overhead of allocating the options and attributes passed to the method so you will need to understand a user can remove that overhead and these numbers will be less". How much less? Developers will need to look at the bechmark code itself to decipher an answer to this. This is already the case for the existing benchmarks.

Prior to this PR I thought the `Add` method required 2 allocations when called with a single use attribute. It turned out to not be true. One of those allocation is from the attribute slice that is being passed to the method, not the method itself.

Reply:
> You can't compare the results meaningfully by going from no preallocating to preallocating as much as possible.

I agree. This comparison is more theoretical then actual, and I called this out in the reported benchmarks:

> These aren't really apples-to-apples comparisons given the new benchmark is for attributes wrapped in an option and the old benchmarks included allocations for reused slices. The allocation reduction should still be the same though if main were updated to wrap an option (adding an allocation) and the attribute slices were reused (removing an allocation).

Reply:
> If a user cannot preallocate in this way, you see that this change increases the number of allocations.

This change isn't so much about "juking the stats" as it is about isolating the performance measurement of the counter's `Add` method.

If a user's code is not structured in a way where they can reuse the input arguments they will have overhead. That is something they will need to evaluate and possibly deal with in their code. Given this benchmark though, what I can confidently tell the user when they are looking at where overhead is coming from is what the overhead from calling the counter's `Add` method is. They can then track down any additional overhead as they know it is in their code. Conversely, that would not be true.

Imagine if a user came to us with the existing code and is trying to track down 1 allocation their code is making where they in the process call the counter's `Add` method with a single use attribute. What can we say to them? The existing benchmark that part of the code should be making 2 allocations, right? How is it they only have 1?

Removing the benchmark fixture overhead from the reported benchmark allows a clear view into the performance and behavior of the code and allows us better understanding.

Reply:
> Given this benchmark though, what I can confidently tell the user when they are looking at where overhead is coming from is what the overhead from calling the counter's `Add` method is.

I'm not sure this is even true. The allocation from this test is from `attribute.NewSet`, not from the counter's `Add` method. I'm wondering if these benchmarks for the "single use" attributes should be removed. They only now measure the setup of attributes being passed to `Add` and no longer have significant representation of the performance of `Add`.

I could add additional benchmarks to `otel/metric/instrument` that measure the performance of the options and configuration creation functions separately. @MadVikingGod what do you think?

Reply:
I agree that the different Add benchmarks all measure essentially the same thing after this change.  What I think is now missing is comparing how users will use this before and after.  How does this change impact someone if they were just using a bag of attributes per Add, and now use `WithAttributes` or create a set and use `WithAttributeSet`.  Essentially the real impacts that a user might experience.  

It could be insightful to have an example optimized like this to show what is needed to get the best performance, but that isn't the data we need to make informed decisions about changes.

Reply:
> How does this change impact someone if they were just using a bag of attributes per Add, and now use WithAttributes or create a set and use WithAttributeSet.

Is this a statement in favor of removing these benchmarks and adding benchmarks for the options?

Reply:
Yes, and to put back in their place something representing a user's experience.  

If we are going to change how users interact and recommend them to use a replacement API, how do I compare the performance of the old and the new?  I want to compare `CounterAddSingleUseAttrs` to something that does an Add with an attribute only known within the loop.  Not the most optimized version, but what we are recommending user use instead of the old API.  This will let us know that we did have more allocations and that the runtime was slower. 

The same can be said for anywhere that you have a shared slice for attribute creation to not count the allocations.

Reply:
> This is not a fair comparison; you can't preallocate both your attributes and options.

> This should be testing the more obvious way people will use it. If you want to include this as an additional optimized benchmark, then I think that would be fair.

I see it differently. First of all, we want to benchmark `WithAttributes` implementation and not how people may (or may not) use it. People should be aware that passing a variadic argument adds allocations - we should not benchmark it. Same for `WithAttributeSet` - we do not want to benchmark the creation of a set that is passed to the function - we have other benchmarks for testing the performance of creating a set.

Of course we might have benchmarks which tests the performance of preparing the args and invoking the functions, but these would be a superset of benchmarks that should only test the function. We could add this, but I would prefer to have it as a separate PR as this one is already very big and I do not see it as a blocker.

Reply:
> Yes, and to put back in their place something representing a user's experience.
> 
> If we are going to change how users interact and recommend them to use a replacement API, how do I compare the performance of the old and the new? I want to compare `CounterAddSingleUseAttrs` to something that does an Add with an attribute only known within the loop. Not the most optimized version, but what we are recommending user use instead of the old API. This will let us know that we did have more allocations and that the runtime was slower.
> 
> The same can be said for anywhere that you have a shared slice for attribute creation to not count the allocations.

It doesn't sound like we are on the same page still. I was planning to remove all the single use stuff. It does not benchmark the performance of a counter's `Add` Method.

It does not make sense to include performance overhead of a benchmark fixture in the benchmark performance. Should we add a sleep to the benchmark as well to simulate the user's common use case where they will be computing things along side calling `Add`? Should we add a few `new([1024]byte)` to also benchmark how users will commonly also make allocations in their code alongside `Add`?

Joking aside, I'm not comfortable keeping benchmarks that are misleading. If you would like to understand the performance impacts of your code I expect you will want to benchmark performance there. I prefer to keep benchmarks in this code base as much about this code base as possible.

Reply:
Additionally, saying we need to include benchmarks for a hypothetical "use case" seems like a fools errand. There is just as strong of a subjective argument to be made that if we didn't preallocate the slices it would not be reflective of the general use case of user. I know that all the instrumentation I manage will reuse arguments as much as possible when calling `Add`. Why is my use case more or less valid to be included here than a user that will not do this?

Measuring expected "general use" is orthogonal to the goal of the benchmark: measuring quantitative performance about this code.

Reply:
I've proposed an update to the base benchmarks [here](https://github.com/open-telemetry/opentelemetry-go/pull/3992).

I'll open an issue to track adding benchmarks for the options and configuration creation functions in `otel/metric`.

Reply:
> I'll open an issue to track adding benchmarks for the options and configuration creation functions in `otel/metric`.

#4002

Reply:
> I've proposed an update to the base benchmarks [here](https://github.com/open-telemetry/opentelemetry-go/pull/3992).

With that merged and this PR updated to follow this comment has become stale. Resolving.